driver/transform-dsl.md

```diff
@@ -0,0 +1,549 @@
+    "right": 5
+}
+```
+
+
+## Restrictions
+
+DSL also enforces few restrictions on transformations and operations.
+
+### Completeness
+
+All operations should either:
+
+* Reject the tree if it does not match one of constraints.
+
+* Describe the whole tree, and save unknown/placeholder nodes to the state.
+
+This rule ensures that if an operation matches the tree, it would be able
+to replicate it from a state associated with the node, since all fields
+are covered by it.
+
+This property allows to test each transformation by executing
+`Check`-`Construct` routine, or detect if a mapping defined by a driver
+developer ignores certain AST fields that are seen in the wild
+(will result in an error).
+
+### Reversibility
+
+Since all operations are required to implement both `Check` and `Construct`
+they can be used interchangeably as either source and/or target transforms.
+
+In simple case it means that transform A can produce the state S for AST
+node T (with `Check`), and then reconstruct an original node T (with `Construct`)
+by only using the state S.
+
+```
+S := A.Check(T); A.Construct(S) == T
+```
+
+In a more complete example, it's possible for transform A to check
+the tree T, produce the state S and for transform B to use this state
+to construct a new tree T'. At the same time the opposite is true:
+transform B can be used to produce state S from T' for A to reconstruct
+original tree T.
+
+```
+// forward transform
+S  := A.Check(T)
+T' := B.Construct(S)
+
+// reverse transform
+S' := B.Check(T')
+S == S'
+T == A.Construct(S')
+```
+
+## List of operations
+
+### Values
+
+#### Any value
+
+In case any value can match the transformation, it should be stored into
+a named variable: `Var("x")`. Any name can be used for the variable as
+long as it is unique in this transformation.
+
+When writing mappings, variable with the same name should exist in the
+second transformation.
+
+#### Equal
+
+To check value for exact match, `Is(v)` operation is used.
+
+SDK defines helpers for common types: `String(v)`, `Int(v)`, etc.
+
+Null values can be checked with `Is(nil)`.
+
+#### Equal or nil
+
+To check any node for nil or specific rule `Opt` is used:
+
+```
+Opt("is_nil", Is(42))
+```
+
+`Opt` needs a named variable to store it's state ("is_nil" in this case).
+
+#### Value in
+
+To check that value equal any value in the list, two operations should be
+executed:
+
+1) Check that value is in the list.
+
+2) Store the value into a variable.
+
+To express this, `Check` helper can be used:
+
+```golang
+Check(
+    // condition
+    In(
+        uast.String("A"),
+        uast.String("B"),
+    ),
+
+    // action
+    Var("x"),
+)
+```
+
+#### Not equal, not in, etc
+
+To check value is not equal `Check` can be combined with `Not`:
+
+```golang
+Check(
+    // condition
+    Not(String("A")),
+
+    // store the value
+    Var("x"),
+)
+```
+
+Note that value can still be of any type and type constraint of `String`
```

Comment:
that **the** value... also the part after and "and" is  a little messy, maybe:

 "Note that the value of any type since we're using `Var` and thus the type constraint of `String` inside the `Check` will not affect the stored variable".